File: clients/kotlin/src/main/kotlin/com/priostack/acn/Results.kt
/*
 * Typed result objects returned by the high-level [AcnClient] methods.
 *
 * The server returns plain JSON. These lightweight data classes give ergonomic,
 * type-checked access to the common calls while still exposing the full raw
 * `data` object via `raw` for anything not surfaced as a property.
 */
package com.priostack.acn

import kotlinx.serialization.json.JsonObject

/** Result of [AcnClient.register]. */
data class RegisterResult(
    val token: String,
    val agentId: String,
    val accountId: String,
    val tier: String,
    val raw: JsonObject,
)

/**
 * Result of [AcnClient.connect].
 *
 * Note the server serialises this with PascalCase Go field names; the client
 * maps them to idiomatic Kotlin properties (`SessionID` -> [sessionId], etc.).
 */
data class ConnectResult(
    val sessionId: String,
    val resolvedAccount: String,
    val grantedCapabilities: List<String>,
    val grantedContextRights: List<String>,
    val boundSpace: String,
    val resolvedMaxTokens: Int,
    val raw: JsonObject,
)

/** Result of [AcnClient.createSpace]. */
data class SpaceResult(
    val spaceId: String,
    val accountId: String,
    val persistenceMode: String,
    val protectionLevel: String,
    val raw: JsonObject,
)

/** Result of [AcnClient.store]. */
data class StoreResult(
    val objectRefs: List<JsonObject>,
    val raw: JsonObject,
) {
    /** How many objects were ingested. */
    val count: Int get() = objectRefs.size
}

/** Result of [AcnClient.fetch]. */
data class FetchResult(
    val space: String,
    val objects: List<JsonObject>,
    val matched: Int,
    val total: Int,
    val returnedTokens: Int,
    val raw: JsonObject,
) {
    /** Just the `content` strings of the returned objects. */
    fun contents(): List<String> = objects.map { it.stringOr("content") }
}

/** Result of [AcnClient.grantAccess] / [AcnClient.approveRequest]. */
data class GrantResult(
    val capabilityRef: String,
    val subject: String,
    val resource: String,
    val effectiveRights: List<String>,
    val raw: JsonObject,
)

/**
 * The `type` of a stored object. The ACN accepts exactly these three kinds;
 * anything else is rejected server-side (proposals/inferences go through the
 * propose/commit path, not `store`).
 */
enum class ObjectType(val wire: String) {
    DECLARATION("declaration"),
    OBSERVATION("observation"),
    MEASUREMENT("measurement"),
}

/**
 * A typed fact to persist with [AcnClient.store].
 *
 * @property content the fact text.
 * @property type one of [ObjectType] (defaults to [ObjectType.DECLARATION]).
 * @property provenance optional source references for the fact.
 */
data class StoreObject(
    val content: String,
    val type: ObjectType = ObjectType.DECLARATION,
    val provenance: List<String> = emptyList(),
)

File: clients/scala/src/main/scala/com/priostack/acn/Results.scala
package com.priostack.acn

/** Typed result objects.
  *
  * The server returns plain JSON. These lightweight case classes give ergonomic,
  * type-checked access to the common calls while still exposing the full raw
  * `data` value via `raw` for anything not surfaced as a field.
  */

/** Result of [[AcnClient.register]]. */
final case class RegisterResult(
    token: String,
    agentId: String,
    accountId: String,
    tier: String,
    raw: ujson.Value = ujson.Null
)

/** Result of [[AcnClient.connect]]. Note the server uses PascalCase Go keys. */
final case class ConnectResult(
    sessionId: String,
    resolvedAccount: String,
    grantedCapabilities: Seq[String],
    grantedContextRights: Seq[String],
    boundSpace: String,
    resolvedMaxTokens: Int,
    raw: ujson.Value = ujson.Null
)

/** Result of [[AcnClient.createSpace]]. */
final case class SpaceResult(
    spaceId: String,
    accountId: String,
    persistenceMode: String,
    protectionLevel: String,
    raw: ujson.Value = ujson.Null
)

/** Result of [[AcnClient.store]]. */
final case class StoreResult(objectRefs: Seq[ujson.Value], raw: ujson.Value = ujson.Null):
  /** How many objects were ingested. */
  def count: Int = objectRefs.size

/** Result of [[AcnClient.fetch]] — the content reader. */
final case class FetchResult(
    space: String,
    objects: Seq[ujson.Value],
    matched: Int,
    total: Int,
    returnedTokens: Int,
    raw: ujson.Value = ujson.Null
):
  /** Just the `content` strings of the returned objects. */
  def contents: Seq[String] =
    objects.map(o => o.objOpt.flatMap(_.get("content")).flatMap(_.strOpt).getOrElse(""))

/** Result of [[AcnClient.grantAccess]] / [[AcnClient.approveRequest]]. */
final case class GrantResult(
    capabilityRef: String,
    subject: String,
    resource: String,
    effectiveRights: Seq[String],
    raw: ujson.Value = ujson.Null
)

/** A typed fact to persist with [[AcnClient.store]].
  *
  * `kind` is the object `type` and must be one of [[AcnClient.StoreKinds]]
  * (`declaration` / `observation` / `measurement`); anything else is rejected
  * client-side before the request is sent.
  */
final case class StoreObject(content: String, kind: String = "declaration", provenance: Seq[String] = Seq.empty)

object StoreObject:
  /** A rule or hypothesis. */
  def declaration(content: String): StoreObject = StoreObject(content, "declaration")
  /** A measured/observed fact. */
  def observation(content: String): StoreObject = StoreObject(content, "observation")
  /** A quantitative measurement. */
  def measurement(content: String): StoreObject = StoreObject(content, "measurement")
